File: leetcode/python/0020_valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stk = []
        for p in s:
            if p in "([{": stk.append(p)
            elif p == ")" and stk and stk[-1] == "(": stk.pop()
            elif p == "]" and stk and stk[-1] == "[": stk.pop()
            elif p == "}" and stk and stk[-1] == "{": stk.pop()
            else: return False
        return len(stk) == 0

File: leetcode/java/0020_valid_parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stk = new Stack<Character>();

        if (s.length() == 1) return false;

        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stk.push(c);
                continue;
            }
            if ((c == ')' || c == '}' || c == ']') && stk.isEmpty()) {
                return false;
            } else if ((c == ')' && stk.peek() != '(') || (c == '}' && stk.peek() != '{') || (c == ']' && stk.peek() != '[')) {
                return false;
            } else {
                stk.pop();
            }
        }

        return stk.isEmpty();
    }
}